File: framework/server/static-assets.js
import fs from 'node:fs';
import fsPromises from 'node:fs/promises';
import path from 'node:path';
import zlib from 'node:zlib';

const MIME_TYPES = {
    '.html': 'text/html',
    '.js': 'application/javascript',
    '.mjs': 'application/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2',
    '.wasm': 'application/wasm'
};

export async function resolveStaticFile(distDir, pathname) {
    if (!pathname) pathname = '/';

    const cleanPathname = pathname.split('?')[0].split('#')[0];
    let normalizedPath = cleanPathname || '/';

    if (normalizedPath.length > 1 && normalizedPath.endsWith('/')) {
        normalizedPath = normalizedPath.slice(0, -1);
    }

    try {
        normalizedPath = decodeURIComponent(normalizedPath);
    } catch (err) {
        return null;
    }

    const absoluteDistDir = path.resolve(distDir);
    const requestedPath = path.join(absoluteDistDir, normalizedPath);

    if (!requestedPath.startsWith(absoluteDistDir + path.sep) && requestedPath !== absoluteDistDir) {
        return null;
    }

    try {
        const stats = await fsPromises.stat(requestedPath);
        if (stats.isFile()) {
            return requestedPath;
        }
    } catch (err) {
        return null;
    }

    return null;
}

export async function serveStaticFile(resolvedPath, request, response) {
    const ext = path.extname(resolvedPath).toLowerCase();
    const contentType = MIME_TYPES[ext] || 'application/octet-stream';

    const filename = path.basename(resolvedPath);
    const isHashedAsset = /\.[a-f0-9]{8,}\.[a-z0-9]+$/i.test(filename);

    const cacheControl = isHashedAsset
        ? 'public, max-age=31536000, immutable'
        : 'public, max-age=0, must-revalidate';

    const headers = {
        'Content-Type': contentType,
        'Cache-Control': cacheControl
    };

    const acceptEncoding = request?.headers?.['accept-encoding'] || '';
    let encoding = null;

    if (acceptEncoding.includes('br')) {
        encoding = 'br';
    } else if (acceptEncoding.includes('gzip')) {
        encoding = 'gzip';
    }

    let streamPipeline = null;

    if (encoding) {
        headers['Vary'] = 'Accept-Encoding';
        const precompressedExt = encoding === 'br' ? '.br' : '.gz';
        const precompressedPath = `${resolvedPath}${precompressedExt}`;

        try {
            const stats = await fsPromises.stat(precompressedPath);
            if (stats.isFile()) {
                headers['Content-Encoding'] = encoding;
                streamPipeline = fs.createReadStream(precompressedPath);
            }
        } catch (err) {
        }

        if (!streamPipeline) {
            headers['Content-Encoding'] = encoding;
            const compressor = encoding === 'br'
                ? zlib.createBrotliCompress()
                : zlib.createGzip();

            streamPipeline = fs.createReadStream(resolvedPath).pipe(compressor);
        }
    } else {
        streamPipeline = fs.createReadStream(resolvedPath);
    }

    return new Promise((resolve, reject) => {
        streamPipeline.on('error', (err) => {
            console.error(`Error serving file ${resolvedPath}:`, err);
            if (!response.headersSent && (!response.statusCode || response.statusCode !== 500)) {
                response.writeHead(500, { 'Content-Type': 'text/plain' });
                response.end('Internal Server Error');
            }
            resolve();
        });

        response.on('finish', resolve);
        response.on('error', reject);

        response.writeHead(200, headers);
        streamPipeline.pipe(response);
    });
}

File: framework/engine-go/static-assets.go
package engine

import (
	"compress/gzip"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var hashedAssetRE = regexp.MustCompile(`[.\-][a-f0-9]{8,}\.[a-zA-Z0-9]+$`)

var mimeTypes = map[string]string{
	".html":  "text/html; charset=utf-8",
	".js":    "text/javascript; charset=utf-8",
	".mjs":   "text/javascript; charset=utf-8",
	".css":   "text/css; charset=utf-8",
	".json":  "application/json; charset=utf-8",
	".svg":   "image/svg+xml",
	".png":   "image/png",
	".jpg":   "image/jpeg",
	".jpeg":  "image/jpeg",
	".gif":   "image/gif",
	".webp":  "image/webp",
	".ico":   "image/x-icon",
	".woff":  "font/woff",
	".woff2": "font/woff2",
	".txt":   "text/plain; charset=utf-8",
	".map":   "application/json; charset=utf-8",
	".wasm":  "application/wasm",
}

func contentTypeFor(path string) string {
	if ct, ok := mimeTypes[strings.ToLower(filepath.Ext(path))]; ok {
		return ct
	}
	return "application/octet-stream"
}

func cacheControlFor(path string) string {
	if hashedAssetRE.MatchString(path) {
		return "public, max-age=31536000, immutable"
	}
	return "public, max-age=0, must-revalidate"
}

func pickEncoding(acceptEncoding string) string {
	lower := strings.ToLower(acceptEncoding)
	if strings.Contains(lower, "br") {
		return "br"
	}
	if strings.Contains(lower, "gzip") {
		return "gzip"
	}
	return ""
}

func Resolve(distDir, pathname string) (path string, ok bool) {
	if idx := strings.IndexByte(pathname, '?'); idx != -1 {
		pathname = pathname[:idx]
	}

	resolvedRoot, err := filepath.Abs(distDir)
	if err != nil {
		return "", false
	}
	candidate := filepath.Join(resolvedRoot, filepath.Clean("/"+pathname))
	if candidate != resolvedRoot && !strings.HasPrefix(candidate, resolvedRoot+string(filepath.Separator)) {
		return "", false
	}

	info, err := os.Stat(candidate)
	if err != nil || info.IsDir() {
		return "", false
	}
	return candidate, true
}

func Serve(w http.ResponseWriter, r *http.Request, filePath string) error {
	encoding := pickEncoding(r.Header.Get("Accept-Encoding"))
	contentType := contentTypeFor(filePath)
	cacheControl := cacheControlFor(filePath)

	if encoding != "" {
		ext := ".gz"
		if encoding == "br" {
			ext = ".br"
		}
		if f, err := os.Open(filePath + ext); err == nil {
			defer f.Close()
			w.Header().Set("Content-Type", contentType)
			w.Header().Set("Cache-Control", cacheControl)
			w.Header().Set("Content-Encoding", encoding)
			w.Header().Set("Vary", "Accept-Encoding")
			w.WriteHeader(http.StatusOK)
			_, err := io.Copy(w, f)
			return err
		}
	}

	f, err := os.Open(filePath)
	if err != nil {
		return err
	}
	defer f.Close()

	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Cache-Control", cacheControl)

	if encoding == "gzip" {
		w.Header().Set("Content-Encoding", "gzip")
		w.Header().Set("Vary", "Accept-Encoding")
		w.WriteHeader(http.StatusOK)
		gz := gzip.NewWriter(w)
		defer gz.Close()
		_, err := io.Copy(gz, f)
		return err
	}

	w.WriteHeader(http.StatusOK)
	_, err = io.Copy(w, f)
	return err
}
